### PR TITLE
Read evdev mappings from file

### PR DESCRIPTION
**Pull request by Holzhaus**
This is needs PRs #752, #764 and #767, so it's actually starting with commit ~~a69f85b~~ (**EDIT:** I rebased, it starts with ee6fca7).

This is somehow big nonetheless, although there are not that many code changed. The biggest thing is the new `core/linux-dist/evdev_keycodes.cpp` (660 lines), so that uses can do `btn_a = BTN_A` instead of `btn_a = 0x130` or `btn_a = 304`. It sucks, I know, but I think it's worth it.

### CODE REVIEW

**Comment by randomstuff** [comment]
It seems [libevdev_event_code_from_name](http://www.freedesktop.org/software/libevdev/doc/1.4.3/group__misc.html#ga6620301a67f467489e4a7f93afe81621) can used to map event name to event code (instead of evdev_keycodes.cpp).

**Comment by randomstuff** [comment]
I'll test this soon. Looks nice! :yum:

**Comment by Holzhaus** [comment]
@randomstuff
> It seems libevdev_event_code_from_name can used to map event name to event code (instead of evdev_keycodes.cpp).

OMG, that'd be awesome. I even asked on stack overflow and nobody knew that. I need to ask @skmp if we want that extra dependency though.

**Comment by randomstuff** [comment]
Yes indeed. I thought there was already a dependency on libevdev but it was on another project…

**Comment by skmp** [comment]
Can you add it as a soft dependency? then TARGET_WITH_LIBEVDEV ->
HOST_HAS_LIBEVDEV (YES/NO). Or maybe link on it on the runtime?

On Tue, Aug 18, 2015 at 11:16 PM, Gabriel Corona <notifications@github.com>
wrote:

> Yes indeed. I thought there was already a dependency on libevdev but it
> was on another project…
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/reicast/reicast-emulator/pull/769#issuecomment-132354407>
> .
>



-- 
~skmp

**Comment by Holzhaus** [comment]
@randomstuff You might want to check this out, i added autoconversion of axis values.

**Comment by randomstuff** [comment]
> added autoconversion of axis values

Great! I'll test all of this tomorrow.

File: shell/linux/mappings/controller_generic.cfg [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+axis_trigger_left = ABS_Z
+axis_trigger_right = ABS_RZ
+
+[compat]
+btn_trigger_left = BTN_TL
+btn_trigger_right = BTN_TR
+axis_dpad_x = ABS_HAT0X
+axis_dpad_y = ABS_HAT0Y
+axis_dpad2_x = ABS_HAT1X
+axis_dpad2_y = ABS_HAT1Y
```
Comment (randomstuff) [inline shell/linux/mappings/controller_generic.cfg]: There's a missing end-of-line here and in some other files.

**Comment by randomstuff** [comment]
I tested this and found no issue. The only remaining issue I have is related to the fact that the values of my analog triggers are inverted.

I have issues when I try to use both analog trigger and the compat trigger buttons: them become both unusable or locked somehow. But it's probably related to the fact that my analog triggers value are inverted.

**Comment by Holzhaus** [comment]
@randomstuff Have you seen 942b259?

**Comment by randomstuff** [comment]
No, I didn't see the commit :) It's working great \o/

One issue is that btn_trigger_left does not seem to be working but axis_trigger_left is working. The configuration seems to be correct. btn_trigger_right is working.

**Comment by Holzhaus** [comment]
Can you post your startup log?

**Comment by randomstuff** [comment]
Sure:

<pre>
Home dir is: /home/test/.reicast/
Personality: 00000000
Updated personality: 00000000
Linux paging: 00001000 00001000 00000FFF
MAP 00800000 w/ 25165824
MAP 20000000 w/ 25165824
MAP 04000000 w/ 16777216
MAP 06000000 w/ 16777216
MAP 0C000000 w/ 0
A8
vmem reserve: base: EE380000, aram: 0e380000, vram: F2380000, ram: FA380000
Resetting mem
Mem alloc successful!Using '/home/test/CrazyTaxi2.cdi' as cd image

libwebsockets test server - (C) Copyright 2010-2013 Andy Green <andy@warmcat.com> - licensed under LGPL2.1
Initial logging level 7
Library version: 1.3 c11b847
IPV6 not compiled in
libev support not compiled in
 static allocation: 4488 + (16 x 65536 fds) = 1053064 bytes
 canonical_hostname = marvin
 per-conn mem: 184 + 1554 headers + protocol rx buf
 Listening on port 5678
Malformed entry on config - ignoring @ 38(ElfFile)
Found 10 matching FB configs.
Chosen visual ID = 0x24
Loaded /home/test/.reicast/data/dc_bios.bin as bootrom

Loaded /home/test/.reicast/data/dc_flash.bin as nvram

Using Recompiler
Auto-selected audio backend "alsa" (Advanced Linux Sound Architecture).
Initializing audio backend "alsa" (Advanced Linux Sound Architecture)...
recSh4 Init
Freeing fpcb

         CodeCache addr: 0x4a2000 | from: 0x4a2000 | addr here: 0x4a0630

Session 1 has 1 track(s)
Saving  Track:  1  Type: Audio/2352  Pregap: 150  Size: 302     LBA: 0       
Track position: 352800

Session 2 has 1 track(s)
Saving  Track:  2  Type: Mode2/2336  Pregap: 150  Size: 308874  LBA: 11702   
Track position: 1413504
gdrom: Opened image "/home/test/CrazyTaxi2.cdi"
Initialising empty vmu...
Initialising empty vmu...
Initialising empty vmu...
Initialising empty vmu...
Initialising empty vmu...
Initialising empty vmu...
Initialising empty vmu...
Sh4 Reset
evdev: Trying to open device at '/dev/input/event17'
evdev: Found 'Thrustmaster Dual Trigger 3-in-1' at '/dev/input/event17'
evdev: reading mapping file: 'controller_generic.cfg'
Malformed entry on config - ignoring @ 14(dpad_left)
Malformed entry on config - ignoring @ 15(dpad_right)
Malformed entry on config - ignoring @ 16(dpad_up)
Malformed entry on config - ignoring @ 17(dpad_down)
Malformed entry on config - ignoring @ 30(axis_dpad2_x)
Malformed entry on config - ignoring @ 31(axis_dpad2_y)
btn_a = BTN_EAST (305)
btn_b = BTN_C (306)
btn_c = BTN_WEST (308)
btn_d = BTN_Z (309)
btn_x = BTN_SOUTH (304)
btn_y = BTN_NORTH (307)
btn_z = BTN_TL (310)
btn_start = BTN_TR2 (313)
btn_escape = -1
dpad_left = -1
dpad_right = -1
dpad_up = -1
dpad_down = -1
dpad2_left = -1
dpad2_right = -1
dpad2_up = -1
dpad2_down = -1
btn_trigger_left = BTN_TL (310)
btn_trigger_right = BTN_TR (311)
axis_dpad_x = ABS_HAT0X (16)
axis_dpad_y = ABS_HAT0Y (17)
axis_dpad2_x = -1
axis_dpad2_y = -1
axis_x = ABS_X (0)
axis_y = ABS_Y (1)
axis_trigger_left = ABS_RX (3)
axis_trigger_right = ABS_RY (4)
evdev: Using 'Generic Controller' mapping
evdev: range of axis 0 is from 0 to 255
evdev: range of axis 1 is from 0 to 255
evdev: range of axis 3 is from 0 to 255
evdev: range of axis 4 is from 0 to 255
evdev: Trying to open device at '/dev/input/event18'
GL context version: 3.1
OSD: 6
Error opening /home/test/.reicast/data/buttons.png
evdev: Found 'Thrustmaster Dual Trigger 3-in-1' at '/dev/input/event18'
evdev: Using 'Generic Controller' mapping
evdev: range of axis 0 is from 0 to 255
evdev: range of axis 1 is from 0 to 255
evdev: range of axis 3 is from 0 to 255
evdev: range of axis 4 is from 0 to 255
evdev: Controller 3 disabled by config.
evdev: Controller 4 disabled by config.
joystick input disabled by config.
</pre>

**Comment by Holzhaus** [comment]
Value is `1` if pressed and `0` if not? I actually can't find an error except that there are malformed entries in the mapping file (look unrelated). I don't have the slightest idea.

**Comment by randomstuff** [comment]
Yes, the value is correct. I'll test again and try to see what's the issue. Thanks for the great work!

**Comment by Holzhaus** [comment]
Btw, I rebased and added a python script to create new mappings. Can somebody test, so that we can merge this after PR #767?

**Comment by skmp** [comment]
What does the python script do?
Am 22.08.2015 02:52 schrieb "Jan Holthuis" <notifications@github.com>:

**Comment by Holzhaus** [comment]
Shows you a list of evdev devices and lets you choose one. It then creates a mapping for it ("press button A", etc).

**Comment by Holzhaus** [comment]
@skmp Here's an example:
http://pastebin.com/LqyvqMP1

**Comment by Holzhaus** [comment]
I'll rebase this evening. Since all needed PRs have been merged, this can go in soon. Please test the python-script to create mapping files though :)

**Comment by Holzhaus** [comment]
Rebased.

**Comment by randomstuff** [comment]
I tested the pytyhon script and it seems to be working great \o/ but the inverted options are not in the correct section.

The C code:

    mf.get_bool("dreamcast", "axis_x_inverted", false),
    mf.get_bool("dreamcast", "axis_y_inverted", false),
    mf.get_bool("dreamcast", "axis_trigger_left_inverted", false),
    mf.get_bool("dreamcast", "axis_trigger_right_inverted", false)

The Python code:

    mapping.set("dreamcast", axisname, event.code)
    mapping.set("compat", "%s_inverted" % axisname, "yes" if axis_inverted else "no")

**Comment by Holzhaus** [comment]
Yikes, I'll fix that.

**Comment by Holzhaus** [comment]
Fixed. Can we merge or are there any issues left?

**Comment by Holzhaus** [comment]
OK, I also disabled the legacy joystick API by default (but people can easily enable it via config file). Also, the python tool and the mappings have been added to the `make install` target in the Makefile. If there are no issues left, we can merge now.

**Pull request merged**